lib/internal/process/modules.js

```diff
@@ -1,17 +1,55 @@
 'use strict';
 
 const {
+  setImportModuleDynamicallyCallback,
   setInitializeImportMetaObjectCallback
 } = internalBinding('module_wrap');
 
+const { getURLFromFilePath } = require('internal/url');
+const Loader = require('internal/loader/Loader');
+const path = require('path');
+const { URL } = require('url');
+
+function normalizeReferrerURL(referrer) {
+  if (typeof referrer === 'string' && path.isAbsolute(referrer)) {
+    return getURLFromFilePath(referrer).href;
+  }
+  return new URL(referrer).href;
+}
+
 function initializeImportMetaObject(wrap, meta) {
   meta.url = wrap.url;
 }
 
-function setupModules() {
+let loaderResolve, loaderReject;
```

Comment:
So, elaborating on what I wrote before:
```js
let aResolve, aReject;
exports.loaderPromise = new Promise((resolve, reject) => {
  aResolve = resolve;
  aReject = reject;
});
trackedPromise.then(aResolve, aReject);
```

Can be written as:

```js
let aResolve;
exports.loaderPromise = new Promise((resolve, reject) => {
  aResolve = resolve;
});
aResolve(trackedPromise);
```

Reply:
could also just use createPromise/promiseReject/promiseResolve, no?

Reply:
@devsnek that's technically true but I'm not sure it would help clean the code and that's an internal API - I doubt the cost of the extra closure here is worth it.

Reply:
I haven't looked through this code at all, so take this with a grain of salt, but I recently did a lot of benchmarking on using the Promise constructor vs the internal createPromise/promiseReject/promiseResolve and the latter is definitely quite a bit faster.

But again, I don't know about the usage here, whether this is performance sensitive code, if these functions end up being user-land exposed, etc.

Reply:
i just meant for readability heh
```js
const loaderPromise = exports.loaderPromise = createPromise();
...
promiseResolve(loaderPromise, ESMLoader);
```

Reply:
@devsnek @apapirovski not allocating a closure and it being faster is actually why we have this API to begin with (as well as `util.promisify`). There's a bit of history involved but if it wasn't faster I'd be surprised. That said - this code does I/O and the impact of an extra closure isn't something we should optimize for. 

@devsnek I think it makes the code easier for people more familiar with promises in core to read but overall I personally prefer standard to nonstandard when we have the choice.

Namely, there are subtle differences between `promiseResolve` and `resolve` in the promise constructor which I'd rather avoid in areas that are not performance sensitive. I don't feel strongly about it though.